MudTable: Add CellClass parameter

boukenka opened the pull request:
## Description
When many columns need to be displayed in a table, even using the ```Dense``` parameter, the table is too wide horizontally. This can be a problem with certain screen sizes. This is why I propose the ```UltraDense``` parameter, which makes the display of data in the table even more compact. 

The ```UltraDense``` parameter takes precedence over the ```Dense``` parameter when set to ```true```.

An example has been added in the _TablePager Customization_ section and offers the possibility to compare the usage of ```Dense``` and ```UltraDense``` parameters.

## How Has This Been Tested?
It has been tested visually. With added sorting, fixed header or comparison of the use of the ```Dense``` parameter.

![UltraDense 1](https://github.com/user-attachments/assets/42c05e78-55f1-461b-be5c-1ecb957ba326)
![UltraDense 2](https://github.com/user-attachments/assets/e791d403-070e-4e87-b7e9-4fb3e1c571ee)
![UltraDense 3](https://github.com/user-attachments/assets/c0511ae7-cb96-41a3-a51e-eb761402e0cf)

## Type of Changes
<!-- What type of changes does your code introduce? Put an `x` in only one box that applies best: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (fix or improvement to the website or code docs)


## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [ ] I've added relevant tests.

Commit:
New UltraDense parameter for MudTable

boukenka commented:
I guess some tests are broken due to the new 9.0.200 SDK based on the @ScarletKuro comment in the [Timeseries Charting improvements](https://github.com/MudBlazor/MudBlazor/pull/10865#).

danielchalmers reviewed (changes requested):
We shouldn't add a new type of property as our density system is already [too disjointed](https://github.com/MudBlazor/MudBlazor/issues/8946), but we do have a `Margin` enum for this purpose and I think deprecating `Dense` and replacing it with that could work. Though it would imply zero padding, not lower (allowing the user to add their own).

Commit:
PaddingDensity parameter for Theme and MudTable

boukenka commented:
@danielchalmers I've made a few changes. First, I've undone all previous modifications. I added a ```Density``` object with a ```PaddingStep``` property to the main ```Theme```. I have removed the effect of the ```Dense``` parameter in the MudTable. Now padding is managed by the ```PaddingDensity``` parameter, which can have four values and apply padding to the table, including the ```PaddingStep``` value in the various calculations. Here are two examples. By default, the ```PaddingParameter``` is set to ```Max```. It makes no changes to the default table in that case for the display. Please let me know your thoughts.

danielchalmers commented:
Thanks for testing that - I think we need a more top-down approach with several components, but it would have to be saved for a major release regardless since it's a breaking change.

I talked to @Garderoben and we think the simplest solution would be to use `MudTd.Class` to change the padding, or do it directly through CSS on the app side, example: https://try.mudblazor.com/snippet/maQzYQlqKmSiBrqc.

It would mean adding the same class to each cell, so a new property (e.g. `CellClass`) could be added to make that easier. What do you think?

boukenka commented:
> I think we need a more top-down approach with several components, but it would have to be saved for a major release regardless since it's a breaking change.

Totally agree with you.

Thanks for discussing this with @Garderoben and for the snippet! Please take a look at https://try.mudblazor.com/snippet/caGTOclgBxcsLLHY. A common class is used to define cells for ```MudTd``` and for ```MudTh``` based on your previous example. Would your idea be to have something like below replace all ```Class=“mud-table-small-cell”``` in the example? If so, I think it would avoid a lot of repetition and could be really useful.

```C#
<MudTable CellClass="mud-table-small-cell"...>
```

Commit:
Parameter CellClass for MudTable

boukenka commented:
I've undone all previous modifications. I have added a ```CellClass``` parameter on the ```<MudTable>``` level. It will be applied to ```<MudTh>``` and ```<MudTd>```. Not to the ```<MudTr>```. I have also added an example. Please see below the two examples. What do you think ?

![One](https://github.com/user-attachments/assets/1800e2c6-b4a4-4883-a68d-603fa36111f2)
![Two](https://github.com/user-attachments/assets/635daf45-a3dd-41c8-bb75-f3c873322f9b)

Commit:
Rollback MudTableGroupRow.razor

Commit:
Tests are now passed

Commit:
Category addition for CellClass parameter

henon commented:
+1 for CellClass. Look at https://mudblazor.com/components/datagrid#visual-styling for comparison. Although we might not want all the `...Style` and `...StyleFunc` params which were added before we decided to not add any more Style parameters.

Garderoben commented:
I like it, this is much better!

But lets not forget about DataGrid, i would like to have CellClass in there as well then since its very similar. Or what do you say @henon @danielchalmers

henon commented:
> But lets not forget about DataGrid, i would like to have CellClass in there as well then since its very similar.

DataGrid has it already and in fact has even better styling suppport. I was suggesting the opposite, getting everything that datagrid supports supported in table as well (except the Style variants, we decided not to add any more Style properties and we may even remove them in the future).

boukenka commented:
On a side note, the ```CellClass``` property for the ```MudDataGrid``` is defined in the ```Column``` component. In this PR, the ```CellClass``` is defined at the ```MudTable``` level. 

The advantage to define the ```CellClass``` at the ```MudTable``` level is that is applied everywhere. No need to define the same class in other components to apply the same class to the cells inside the ```MudTable``` component. 

One solution could be to do the following, for consistency :

A - Rename ```CellClass``` and ```CellClassFunc``` for ```MudDataGrid``` component in ```ColumnClass``` and ```ColumnClassFunc```. This is a breaking change.
B - Move ```CellClass``` (and ```CellClassFunc``` ?)  for ```MudDataGrid``` component at the ```MudDataGrid``` level. This is a breaking change as the point A. These properties therefore behave in the same way as this PR.
C - Add ```RowClass```, ```RowClassFunc```, ```HeaderClass```, ```HeaderClassFunc```, ```FooterClass```,  ```FooterClassFunc```, ```ColumnClass``` and ```ColumnClassFunc``` to ```MudTable``` component (if this is possible).
D - Add as well ```CellClassFunc``` to ```MudTable``` component.

What do you think? Or keep this PR as it is? Or another solution ?

henon commented:
MudGrid is column based and MudTable is not. They have a different design philosophy and so I think we're fine with the different behavior of CellClass. So I would not change MudGrid, meaning I'd not do A and B. 

If you want to add C in this PR you are welcome to do so. If you don't have the time then just add CellClassFunc. It should have parameters to identify the row and column of the table so could theoretically have a different style for every cell in the table. I guess the func would also differ from MudGrid because the grid does not need information about the column as a parameter in the func.

Commit:
Parameter CellClassFunc for MudTable

Commit:
Rollback MudTable.razor

boukenka commented:
Here is an updated version of the PR. I am pretty sure there is a better way to implement the ```CellClassFunc``` parameter.

![One](https://github.com/user-attachments/assets/d2a7701b-1f55-47a6-af2f-693d6a9ab0b7)
![Two](https://github.com/user-attachments/assets/c29c3cba-17de-4df4-baf2-7d36c88e056c)

danielchalmers commented:
Will leave this to @henon

ScarletKuro commented on src/MudBlazor/Components/Table/MudTr.razor:
```diff
@@ -36,7 +36,7 @@
     @if (Expandable || Checkable)
     {
-        <MudElement HtmlTag="@("Td")" Class="mud-table-cell">
+        <MudElement HtmlTag="@("Td")" Class="@($"mud-table-cell {Context?.Table?.CellClass}")">
```
@danielchalmers wouldn't be using CssBuilder better here?

boukenka replied:
CssBuilder added.

ScarletKuro commented on src/MudBlazor/Components/Table/MudTable.razor.cs:
```diff
@@ -762,6 +772,33 @@ private string ClearFilterCache()
 
+        private int _cellIndex;
+
+        public override string GetCellClassFuncAsync()
```
Why does this have `Async` prefix?

boukenka replied:
Fixed.

ScarletKuro commented on src/MudBlazor/Components/Table/MudTable.razor.cs:
```diff
@@ -762,6 +772,33 @@ private string ClearFilterCache()
             return "";
         }
 
+        private int _cellIndex;
```
we put the private fields at the start of the source file

boukenka replied:
Fixed

ScarletKuro commented:
Somehow surprised how much code requires for a `CellClassFunc`, I'm guessing it's because in datagrid it's per column, and here it's on the table level. In datagrid it's just simple as:
https://github.com/MudBlazor/MudBlazor/blob/b40ab85516ae8cf2656dfc2022a6fbf1eac505d2/src/MudBlazor/Components/DataGrid/Cell.cs#L32-L33 and no more code

Commit:
Feedback to be resolved

boukenka commented:
> Somehow surprised how much code requires for a `CellClassFunc`, I'm guessing it's because in datagrid it's per column, and here it's on the table level. In datagrid it's just simple as:
> 
> https://github.com/MudBlazor/MudBlazor/blob/b40ab85516ae8cf2656dfc2022a6fbf1eac505d2/src/MudBlazor/Components/DataGrid/Cell.cs#L32-L33
> 
> https://github.com/MudBlazor/MudBlazor/blob/b40ab85516ae8cf2656dfc2022a6fbf1eac505d2/src/MudBlazor/Components/DataGrid/MudDataGrid.razor#L485
> 
> and no more code

Yes, this is much simpler. Please let me know if you see a better implementation for ```CellClassFunc```.

Anu6is commented:
Same code, just minor refactor. The `CalculateCellPosition` may not be necessary but I think the `ShouldApplyCellClass` helps. 

```cs
public override string GetCellClassFunc()
{
    if (!ShouldApplyCellClass())
        return string.Empty;
    
    var cellCount = Context.Table.GetCellsQuantity();
    var (columnIndex, rowIndex) = CalculateCellPosition(cellCount);

    _cellIndex = _cellIndex == cellCount - 1 ? 0 : _cellIndex++;

    return CellClassFunc.Invoke(Items.ElementAt(rowIndex), columnIndex, rowIndex);
}

private bool ShouldApplyCellClass()
{
    return CellClassFunc is not null
        && Context?.Table is not null
        && !Virtualize
        && GroupBy is null
        && Items is not null
        && Context.Table.GetRowsQuantity() != 0
        && !CustomHeader  
        && FooterContent is null  
        && !MultiSelection; 
}

private (int columnIndex, int rowIndex) CalculateCellPosition(int cellCount)
{
    var columnsCount = cellCount / Context.Table.GetRowsQuantity();
    var columnIndex = _cellIndex % columnsCount;
    var rowIndex = _cellIndex / columnsCount;
    
    return (columnIndex, rowIndex);
}
```

Also, why the last 3 conditions in `ShouldApplyCellClass()`? 
Just wondering why isn't the function applied if CustomHeader or FooterContent or MultiSelection?

boukenka commented:
> Same code, just minor refactor. The `CalculateCellPosition` may not be necessary but I think the `ShouldApplyCellClass` helps.
> 
> ```cs
> public override string GetCellClassFunc()
> {
>     if (!ShouldApplyCellClass())
>         return string.Empty;
>     
>     var cellCount = Context.Table.GetCellsQuantity();
>     var (columnIndex, rowIndex) = CalculateCellPosition(cellCount);
> 
>     _cellIndex = _cellIndex == cellCount - 1 ? 0 : _cellIndex++;
> 
>     return CellClassFunc.Invoke(Items.ElementAt(rowIndex), columnIndex, rowIndex);
> }
> 
> private bool ShouldApplyCellClass()
> {
>     return CellClassFunc is not null
>         && Context?.Table is not null
>         && !Virtualize
>         && GroupBy is null
>         && Items is not null
>         && Context.Table.GetRowsQuantity() != 0
>         && !CustomHeader  
>         && FooterContent is null  
>         && !MultiSelection; 
> }
> 
> private (int columnIndex, int rowIndex) CalculateCellPosition(int cellCount)
> {
>     var columnsCount = cellCount / Context.Table.GetRowsQuantity();
>     var columnIndex = _cellIndex % columnsCount;
>     var rowIndex = _cellIndex / columnsCount;
>     
>     return (columnIndex, rowIndex);
> }
> ```

Thanks for the code proposition. But with this code, you get _Dereference of a possibly null reference_ for ```Context.Table``` in ```GetCellClassFunc()``` method and ```CalculateCellPosition()``` method. Also for ```Items``` and ```CellClassFunc``` in ```GetCellClassFunc()``` method. So I am not sure it is good to apply ```!``` to dereference the null value for all these cases even this is true and verified by the ```ShouldApplyCellClass``` method. What do you think?

> Also, why the last 3 conditions in `ShouldApplyCellClass()`? Just wondering why isn't the function applied if CustomHeader or FooterContent or MultiSelection?

Because, for example, for the ```FooterContent,``` with a ```<MudFootRow>```, there may be a ```<MudTd>```. This ```<MudTd>``` will be rendered before the ```<MudTd>``` included in ```<RowTemplate>```. All calculations will then become incorrect for the x and y cell positions. 

This is why I think there must be a better way to implement the ```CellClassFunc``` functionality. For the ```CellClass``` parameter implementation, I think it's ok.

Anu6is commented:
Wrote the code directly in the GitHub comment so definitely clean up whatever needs to be fixed 🙃

Commit:
Rollback to CellClass implementation

boukenka commented:
@Henon @ScarletKuro I've made several attempts to 'correctly' count the number of cells when sorting, navigating to another page or changing the number of rows to display. Each time, the code became more and more complex to get the ```CellClassFunc``` calls right. But I couldn't solve every scenario. So I decided to go back to my original implementation of the ```CellClass``` parameter. Could you review the PR as it is? I'm stuck for a correct solution to implement the ```CellClassFunc``` parameter. Maybe someone will find a good solution to implement this feature.

henon committed:
Update TablePage.razor

henon committed:
Update TablePage.razor

henon reviewed (approved):
LGTM

ScarletKuro committed:
Merge branch 'dev' into compact

ScarletKuro committed:
Update TableCellClassExample.razor
